### PR TITLE
wgengine/{monitor,router}: restore Linux ip rules when systemd deletes them

### LINKED ISSUE #1591: systemd-networkd removes tailscale routing table entries
**Describe the bug**

If during the boot the networking/DNS is not available yet at the moment tailscaled starts, the tailscale interface never comes up and stucks in a loop like:

```
Mar 26 12:25:01 rain tailscaled[283]: logtail: dial "log.tailscale.io:443" failed: dial tcp: lookup log.tailscale.io: Temporary failure in name resolution (in 1ms)
Mar 26 12:25:01 rain tailscaled[283]: logtail: upload: log upload of 307 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/…": dial tcp: lookup log.tailscale.io: Temporary failure in name resolution
Mar 26 12:25:01 rain tailscaled[283]: logtail: backoff: 39 msec
…
Mar 26 12:26:14 rain tailscaled[404]: trying bootstrapDNS("derp3.tailscale.com", "68.183.179.66") for "login.tailscale.com" ...
Mar 26 12:26:14 rain tailscaled[404]: bootstrapDNS("derp3.tailscale.com", "68.183.179.66") for "login.tailscale.com" error: Get "https://derp3.tailscale.com/bootstrap-dns?q=login.tailscale.com": dial tcp 68.183.179.66:443: connect: network is unreachable
…
Mar 26 12:26:14 rain tailscaled[404]: control: authRoutine: backoff: 10588 msec
Mar 26 12:26:18 rain tailscaled[404]: logtail: dial "log.tailscale.io:443" failed: dial tcp 34.210.105.16:443: connect: network is unreachable (in 3ms)
Mar 26 12:26:18 rain tailscaled[404]: logtail: upload: log upload of 422 bytes compressed failed: Post "https://log.tailscale.io/c/tailnode.log.tailscale.io/…": dial tcp 34.210.105.16:443: connect: network is unreachable
Mar 26 12:26:18 rain tailscaled[404]: logtail: backoff: 12570 msec
Mar 26 12:26:24 rain tailscaled[404]: control: authRoutine: state:authenticating; wantLoggedIn=true
Mar 26 12:26:24 rain tailscaled[404]: control: direct.TryLogin(token=false, flags=0)
Mar 26 12:26:24 rain tailscaled[404]: control: doLogin(regen=false, hasUrl=false)
```

even if the networking is up and running a little bit later, and required hosts are perfectly available through telnet.

Restarting tailscaled fixes the issue.

**To Reproduce**
1. install tailscaled on Arch Linux
2. in my case, local DNS resolver is used (_systemd-resolved+dnscrypt_)

**Expected behavior**
tailscaled should restore connectivity once networking is up and running.

**Screenshots**
N/A

**Version information:**
 - Device: various machines
 - OS: Arch Linux
 - OS version: latest
 - Tailscale version: v1.6.0

**Additional context**
I'm not using MagicDNS/nameservers/anything else available in the webadmin.

It seems I've managed to work around the issue by overriding unit dependencies:

```ini
Wants=network-online.target
After=network.target network-online.target
```
so that the daemon starts *after* networking/DNS is available.

I'm not sure this is a proper solution though, so please help.

Thanks.

### PR DESCRIPTION
**Pull request by bradfitz**
Thanks.

Fixes #1591

### CODE REVIEW

**Comment by bradfitz** [comment]
Haven't finished testing, but should work from earlier experiments. Will finish testing in a bit here.

**Comment by josharian** [comment]
Is there any danger of getting into a fight with systemd, in which hit an infinite loop of delete/restore?

File: wgengine/router/router_linux.go [hunk 1]
```diff
@@ -164,10 +165,20 @@ func newUserspaceRouterAdvanced(logf logger.Logf, tunname string, linkMon *monit
...
+
+	return r, nil
 }
 
 func (r *linuxRouter) Up() error {
+	if r.unregLinkMon == nil && r.linkMon != nil {
+		r.unregLinkMon = r.linkMon.RegisterRuleDeleteCallback(func(table uint8, priority uint32) {
+			if priority >= 5200 && priority < 5300 {
+				r.logf("somebody (systemd-networkd?) deleted Tailscale's ip rules; restoring")
+				r.addIPRules()
```
Comment (crawshaw) [inline wgengine/router/router_linux.go:178]: is `addIPRules` robust when only some of the rules are deleted?

File: wgengine/router/router_linux.go [hunk 2]
```diff
@@ -164,10 +165,20 @@ func newUserspaceRouterAdvanced(logf logger.Logf, tunname string, linkMon *monit
...
-	}, nil
+	}
+
+	return r, nil
 }
 
 func (r *linuxRouter) Up() error {
+	if r.unregLinkMon == nil && r.linkMon != nil {
+		r.unregLinkMon = r.linkMon.RegisterRuleDeleteCallback(func(table uint8, priority uint32) {
+			if priority >= 5200 && priority < 5300 {
```
Comment (crawshaw) [inline wgengine/router/router_linux.go:176]: not checking the `table`?

**Review by crawshaw — APPROVED** [review]

File: wgengine/router/router_linux.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/router/router_linux.go]: Yes, but its delete itself caused the cleaner to run, so broke it in half with a justAddIPRules part.

File: wgengine/router/router_linux.go (hunk 2 above)
Comment (bradfitz) [reply wgengine/router/router_linux.go]: Doesn't matter. Added a comment about why.

**Comment by bradfitz** [comment]
> Is there any danger of getting into a fight with systemd, in which hit an infinite loop of delete/restore?

I'd feared that but empirically not.

**Comment by bradfitz** [comment]
> Is there any danger of getting into a fight with systemd, in which hit an infinite loop of delete/restore?

But added a rate limiter.

**Comment by danderson** [comment]
If it is in fact systemd-networkd messing with us, when no need to fear tug of war: by design, it only does work when it first starts up, to bring the system into the desired state. It doesn't try to enforce the state beyond that.

**Pull request merged**